scratchpad/qaoa_eng_vs_be_adv.py

```diff
@@ -0,0 +1,197 @@
+import json
+from functools import lru_cache
+from attr import asdict
+from networkx.readwrite.gml import LIST_START_VALUE
+import numpy as np
+import networkx as nx
+import pytest
+import platform
+import pyrofiler
+from qtensor import QtreeQAOAComposer
+from qtensor import QAOAQtreeSimulator
+from qtensor.contraction_backends import get_backend, get_perf_backend
+
+@lru_cache
+def get_test_problem(n=10, p=2, d=3, type='random'):
+    print('Test problem: n, p, d', n, p, d)
+    if type == 'random':
+        G = nx.random_regular_graph(d, n)
+    elif type == 'grid2d':
+        G = nx.grid_2d_graph(n,n)
+    elif type == 'line':
+        G = nx.Graph()
+        G.add_edges_from(zip(range(n-1), range(1, n)))
+    gamma, beta = [np.pi/5]*p, [np.pi/2]*p
+    return G, gamma, beta
+
+@pytest.fixture
+def test_problem(request):
+    n, p, d, type = request.param
+    return get_test_problem(n, p, d, type)
+
+
+paramtest = [
+    # n, p, degree, type
+     [4, 4, 3, 'random']
+    ,[10, 5, 2, 'random']
+    ,[14, 1, 3, 'random']
+    ,[3, 3, 0, 'grid2d']
+    ,[8, 4, 0, 'line']
+]
+
+def mean_mmax(x: list):
+    mx, mn = max(x), min(x)
+    x.remove(mx)
+    x.remove(mn)
+    return np.mean(x)
+
+def format_flops(flops):
+    ord = 3*int(np.log10(flops)/3)
+    suffix = {
+        3: 'k'
+        ,6: 'M'
+        ,9: 'G'
+        , 12: 'T'
+    }[ord]
+    return f'{(flops/10**ord).round(2)}{suffix}'
+
+def obj2dict(obj):
+    keys = [x for x in dir(obj) if x[0]!='_']
+    return dict((key, obj.__getattribute__(key)) for key in keys)
+
+
+@lru_cache
+def get_gpu_props_json():
+    try:
+        import torch
+        devprops = torch.cuda.get_device_properties(torch.cuda.current_device())
+        return obj2dict(devprops)
+    except:
+        return None
+
+'''
+Main Functions:
+'''
+
+'''
+Function: Generate a single report of a backend running a problem
```

Comment:
Move this to a [docstring](https://www.python.org/dev/peps/pep-0257/#what-is-a-docstring). How to write docstrings: https://google.github.io/styleguide/pyguide.html#383-functions-and-methods